pkg/apis/networking/validation/validation.go

```diff
@@ -561,6 +569,51 @@ func validateIngressTypedLocalObjectReference(params *api.TypedLocalObjectRefere
 	return allErrs
 }
 
+// validateIngressClassParametersReference ensures that Parameters fields are valid.
+// Parameters was previously of type `TypedLocalObjectReference` and used
+// `validateIngressTypedLocalObjectReference()`. This function extends validation
+// for additional fields introduced for namespace-scoped references.
+func validateIngressClassParametersReference(params *networking.IngressClassParametersReference, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if params == nil {
+		return allErrs
+	}
+
+	allErrs = append(allErrs, validateIngressTypedLocalObjectReference(&api.TypedLocalObjectReference{
+		APIGroup: params.APIGroup,
+		Kind:     params.Kind,
+		Name:     params.Name,
+	}, fldPath)...)
+
```

Comment:
I find the below block hard to reason about.  Here's what I think it should
look like, at a high level (rough, please cross-check):

```
if gate-enabled && scope == nil {
	// error: required value
} else if scope != nil {
	// if we get here, the gate may or may not be enabled, but scope has a value.
	if not valid scope value
		// error: not supported
	} else {
	    // if we get here, we have a valid scope.
	    if scope == cluster
		    if namespace != nil {
			    // error: forbidden when scope is cluster
		    }
	    }
	    if scope == Namespace {
		    if namespace == nil {
			    // error: required when scope is NS
		    }
	   }
	}
}
```

Then, when we eventually remove the gate, it's simply removing the outermost
if/else.

Does that make sense?

Reply:
Great timing. I was just working on this logic and I arrived close to where you are.
We can't use `else if` because in that case, when gate is enabled a value is set (to garbage), the majority of validation will be skipped.


This is what I've right now (added comments here, not in git patch):
```go
        // when gate goes away, we have to this if block only
        if utilfeature.DefaultFeatureGate.Enabled(features.IngressClassNamespacedParams) && params.Scope == nil {
                allErrs = append(allErrs, field.Required(fldPath.Child("scope"), "scope is required"))
                return allErrs
        }
        // need to check for namespace too because if scope is not set, but namespace is we still validate
        if params.Scope != nil || params.Namespace != nil {
                scope := utilpointer.StringPtrDerefOr(params.Scope, "")
                namespace := utilpointer.StringPtrDerefOr(params.Namespace, "")
 
                if !supportedIngressClassParametersReferenceScopes.Has(scope) {
                        allErrs = append(allErrs, field.NotSupported(fldPath.Child("scope"), scope,
                                supportedIngressClassParametersReferenceScopes.List()))
                } else {

                        if scope == networking.IngressClassParametersReferenceScopeNamespace {
                                if namespace == "" {
                                        allErrs = append(allErrs, field.Required(fldPath.Child("namespace"), "namespace is required when spec.parameters.scope is set to Namespace"))
                                } else {
                                       // performs namespace validation only when needed
                                        for _, msg := range apivalidation.ValidateNamespaceName(namespace, false) {
                                                allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), namespace, msg))
                                        }
                                }
                        }

                        if scope == networking.IngressClassParametersReferenceScopeCluster && namespace != "" {
                                allErrs = append(allErrs, field.Forbidden(fldPath.Child("namespace"), "namespace is forbidden when spec.parameters.scope is set to Cluster"))
                        }
                }
        }
```

How does that sound?